src/Clay/Grid.hs

```diff
@@ -56,3 +109,669 @@ gridGap = key "grid-gap"
 -- | Property defines the line names and track sizing functions of the grid columns.
 gridTemplateColumns :: [Size a] -> Css
 gridTemplateColumns = key "grid-template-columns" . noCommas
+
+-- $sizeAndLocationIntro
+--
+-- == CSS documentation
+-- The below functions are based on
+-- [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/CSS)
+-- CSS documentation.
+--
+-- === __Naming note__
+-- In this documentation, as the functions are polymorphic we sometimes
+-- refer to the CSS types as used in the
+-- [mdm web docs_](https://developer.mozilla.org/en-US/docs/Web/CSS)
+-- rather than the Haskell types.
+-- For example, @grid-line@ is used instead of 'GridLine' as a the argument
+-- might be provided as a 'GridLine' but also as an 'Integer', 'String', etc.
+--
+-- == Pragma
+-- #pragma#
+-- If you want to avoid specifying the types of the arguments, enable
+-- the @ExtendedDefaultRules@ GHC language pragma as well as the
+-- @-Wno-type-defaults@ GHC option to avoid compilation warnings.
+--
+-- @
+-- {-# LANGUAGE ExtendedDefaultRules #-}
+-- {-# OPTIONS_GHC -Wno-type-defaults #-}
+-- @
+--
+-- === __Examples__
+-- With the above enabled, you can write:
+--
+-- >> gridRowStart 2
+--
+-- >> gridRowStart "somegridarea"
+--
+-- If you do not enable those, then you must write:
+--
+-- >> gridRowStart (2 :: Integer)
+--
+-- >> gridRowStart ("somegridarea" :: String)
+--
+-- === __Note__
+-- If you decide to enable the above, it is advisable to have your Clay
+-- CSS code in its own module, so the behaviour of the rest of your code
+-- is not affected.
+--
+-- == Examples
+-- Examples are provided through the documentation for the various functions.
+-- Further examples can be found in the source code of the test suite
+-- in the GridSpec.hs module.
+
+-- | A @grid-line@ value.
+--
+-- A @grid-line@ value specifies a size and location in a grid.
+data GridLine
+
+  -- | 'Integer' value.
+  --
+  -- __NOTE:__ 'Integer' value of 0 is invalid.
+  = Coordinate Integer
+
+  -- | @custom-ident@ with an optional 'Integer' value.
+  --
+  -- __NOTE:__ 'Integer' value of 0 is invalid.
+  | GridLineCustomIdent CustomIdentGrid (Maybe Integer)
+
+  -- | @span@ CSS keyword with an optional @custom-ident@ and/or 'Integer' value.
+  --
+  -- __NOTE:__ negative 'Integer' or 0 are invalid.
+  | Span (Maybe CustomIdentGrid) (Maybe Integer)
+
+  -- | @auto@ CSS keyword.
+  | Auto
+
+  -- | @inherit@ CSS keyword.
+  | Inherit
+
+  -- | @initial@ CSS keyword.
+  | Initial
+
+  -- | @unset@ CSS keyword.
+  | Unset
+  deriving (Eq, Show)
```

Comment:
This is different to other typical implementations of such values. The usual style would be to add a catchall constructor encapsulating a `Value` and implement all typeclasses that the CSS specification allows. The constructors are usually hidden anyways, so there is no problem of being able to encode more than is allowed.

```suggestion
  -- | Other grid line values such as @auto@, @inherit@, ...
  | OtherGridLine Value
  deriving (Eq, Show, Auto, Inherit, Initial, Unset)
```

Reply:
This is a very valid point!

However, the listed keywords are the ones existing as such in the `Clay.Common` module.
I would therefore suggest to keep those, as they ensure type safety for the user. Not all keywords are applicable to grids (for example, you couldn't use `visible`).

Regarding the catchall, we would also need an instance (around line 650) for the constructor, using the class `Other` of the `Clay.Common` module.

Shall I submit a suggestion?